### PR TITLE
Persist ETH1 Data to Disk

### LINKED ISSUE #1430: Store Previously Requested ETH1.0 Logs Block Number
Hi all,

Currently our powchain service reads all historical logs from the deposit contract from ETH1.0. As the contract grows even more, we should be persisting the last requested block number for logs to disk in the powchain service to avoid looping through all of these logs every time the node is launched. a PR for this would store the last requested ETH1.0 block number for log fetching in BoltDB. Next time the node spins up, we can request logs started from this block number + 1. Thanks @nisdas for bringing this up.

### LINKED ISSUE #4370: Genesis validators incorrect when min genesis time in future
Problem scenario:
- Send more than `MinGenesisActiveValidatorCount` valid deposits before `MinGenesisTime`.
- When current time exceeds `MinGenesisTime`, all valid deposits are added to the genesis state.
- New nodes that sync after genesis time have only `MinGenesisActiveValidatorCount` validators.

Expected behavior:
- Same behavior as pre-genesis nodes.

Edit: Edited to update based on feedback from @nisdas.

### PR DESCRIPTION
**Pull request by nisdas**
Builds on #4306 and resolves #1430 and resolves #4370

- [x] Save regular checkpoints of eth1data to disk.
- [x] Be able to start up node with persisted eth1 data and not relying on 
previous deposit logs
- [x] Save powchain data at regular intervals
- [x] Remove log block delay as that was a cause of a genesis bug

### CODE REVIEW

File: beacon-chain/blockchain/service_test.go [hunk 1]
```diff
@@ -9,6 +9,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+
```
Comment (rauljordan) [inline beacon-chain/blockchain/service_test.go]:
```suggestion
```

File: beacon-chain/core/state/state.go [hunk 2]
```diff
@@ -187,6 +187,127 @@ func GenesisBeaconState(deposits []*ethpb.Deposit, genesisTime uint64, eth1Data
 	return state, nil
 }
 
+// OptimizedGenesisBeaconState is used to create a state that has already processed deposits. This is to efficiently
+// create a mainnet state at chainstart.
+func OptimizedGenesisBeaconState(genesisTime uint64, bState *pb.BeaconState, eth1Data *ethpb.Eth1Data) (*pb.BeaconState, error) {
```
Comment (rauljordan) [inline beacon-chain/core/state/state.go]: Is the idea to eventually deprecate the original genesis state func? It's a bit weird to have an optimized and non-optimized version of a function. Perhaps some refactoring could be used so we have only one of these functions and the heavy lifting is done outside of it?

File: beacon-chain/db/iface/interface.go [hunk 3]
```diff
@@ -5,6 +5,8 @@ import (
 	"context"
 	"io"
 
+	"github.com/prysmaticlabs/prysm/proto/beacon/db"
+
```
Comment (rauljordan) [inline beacon-chain/db/iface/interface.go]:
```suggestion
```

File: beacon-chain/db/iface/interface.go [hunk 4]
```diff
@@ -85,4 +87,8 @@ type Database interface {
 	// Deposit contract related handlers.
 	DepositContractAddress(ctx context.Context) ([]byte, error)
 	SaveDepositContractAddress(ctx context.Context, addr common.Address) error
+
```
Comment (rauljordan) [inline beacon-chain/db/iface/interface.go]:
```suggestion
```

File: beacon-chain/db/kafka/passthrough.go [hunk 5]
```diff
@@ -3,6 +3,8 @@ package kafka
 import (
 	"context"
 
+	"github.com/prysmaticlabs/prysm/proto/beacon/db"
+
```
Comment (rauljordan) [inline beacon-chain/db/kafka/passthrough.go]:
```suggestion
```

File: beacon-chain/powchain/deposit.go [hunk 6]
```diff
@@ -1,82 +1,34 @@
...
-		leaf[:],
-		int(eth1Data.DepositCount-1),
-		deposit.Proof,
-	); !ok {
-		return fmt.Errorf(
-			"deposit merkle branch of deposit root did not verify for root: %#x",
-			receiptRoot,
-		)
+	balance := s.preGenesisState.Balances[index]
+	s.preGenesisState.Validators[index].EffectiveBalance = mathutil.Min(balance-balance%params.BeaconConfig().EffectiveBalanceIncrement, params.BeaconConfig().MaxEffectiveBalance)
```
Comment (rauljordan) [inline beacon-chain/powchain/deposit.go]: this is a very long expression. Can these be put in perhaps some small helper funcs in this file? Or maybe some of  of the parts of this expression can be broken up into variables, we also need comprehensive tests for this particular few lines

File: beacon-chain/powchain/deposit_test.go [hunk 7]
```diff
@@ -275,18 +285,22 @@ func TestProcessDeposit_AllDepositedSuccessfully(t *testing.T) {
...
+		}
+
+		if valCount != uint64(i+1) {
+			t.Errorf("Did not get correct active validator count received %d, but wanted %d", valCount, uint64(i+1))
 		}
-		pubkey := bytesutil.ToBytes48(k.PublicKey().Marshal())
-		if web3Service.depositedPubkeys[pubkey] != params.BeaconConfig().MaxEffectiveBalance {
-			t.Errorf("Wanted a full deposit of %d but got %d", params.BeaconConfig().MaxEffectiveBalance, web3Service.depositedPubkeys[pubkey])
+		if web3Service.preGenesisState.Validators[i].EffectiveBalance != params.BeaconConfig().MaxEffectiveBalance {
+			t.Errorf("Wanted a full deposit of %d but got %d", params.BeaconConfig().MaxEffectiveBalance, web3Service.preGenesisState.Validators[i].EffectiveBalance)
```
Comment (rauljordan) [inline beacon-chain/powchain/deposit_test.go]: Also long expression, consider breaking up into a few lines

File: beacon-chain/powchain/log_processing.go [hunk 8]
```diff
@@ -7,6 +7,10 @@ import (
 	"math/big"
 	"time"
 
+	protodb "github.com/prysmaticlabs/prysm/proto/beacon/db"
+
```
Comment (rauljordan) [inline beacon-chain/powchain/log_processing.go]:
```suggestion
```

File: beacon-chain/powchain/log_processing.go [hunk 9]
```diff
@@ -7,6 +7,10 @@ import (
...
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+
```
Comment (rauljordan) [inline beacon-chain/powchain/log_processing.go]:
```suggestion
```

File: beacon-chain/powchain/log_processing.go [hunk 10]
```diff
@@ -73,6 +77,15 @@ func (s *Service) ProcessLog(ctx context.Context, depositLog gethTypes.Log) erro
 		if err := s.ProcessDepositLog(ctx, depositLog); err != nil {
 			return errors.Wrap(err, "Could not process deposit log")
 		}
+		if s.lastReceivedMerkleIndex%100 == 0 {
```
Comment (rauljordan) [inline beacon-chain/powchain/log_processing.go]: what's the 100? can we make this a const at the package level?

File: beacon-chain/powchain/log_processing.go [hunk 11]
```diff
@@ -230,47 +244,62 @@ func (s *Service) ProcessChainStart(genesisTime uint64, eth1BlockHash [32]byte,
 }
 
 func (s *Service) setGenesisTime(timeStamp uint64) {
+	s.chainStartData.GenesisTime = s.createGenesisTime(timeStamp)
```
Comment (rauljordan) [inline beacon-chain/powchain/log_processing.go]: do we really need this one liner to be a separate func?

File: beacon-chain/powchain/testing/mock.go [hunk 12]
```diff
@@ -6,6 +6,8 @@ import (
 	"math/big"
 	"time"
 
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+
```
Comment (rauljordan) [inline beacon-chain/powchain/testing/mock.go]:
```suggestion
```

File: shared/trieutil/sparse_merkle.go [hunk 13]
```diff
@@ -6,6 +6,8 @@ import (
 	"errors"
 	"fmt"
 
+	protodb "github.com/prysmaticlabs/prysm/proto/beacon/db"
+
```
Comment (rauljordan) [inline shared/trieutil/sparse_merkle.go]:
```suggestion
```

File: beacon-chain/powchain/deposit_test.go [hunk 14]
```diff
@@ -5,10 +5,11 @@ import (
 	"strings"
 	"testing"
 
+	testDB "github.com/prysmaticlabs/prysm/beacon-chain/db/testing"
+
```
Comment (prestonvanloon) [inline beacon-chain/powchain/deposit_test.go]: remove space / gofmt pls

**Pull request merged**